### PR TITLE
feat: Implement alternate config lookup

### LINKED ISSUE #18385: Change Request: Make it easier to inherit flat configs from the repo root
### ESLint version

v8.57.0

### What problem do you want to solve?

When using a flat config, the CLI only ever loads one config; this is resolved depending on where you invoke it. In a monorepo, you may want to have multiple `eslint.config.js` files; this is common for non-flat configs, where child `.eslintrc.json` files merge with their parents.

It's a design choice that flat configs no longer merge with parent directories; this is fine, and it seems reasonable that one could write a config like:

```js
import rootConfig from "../../../eslint.config.js";

export default [
    ...rootConfig,
    // Some more stuff
]
```

However, this doesn't acutally work!

- If you run `eslint .` at the root of the monorepo, eslint only loads the root config. All other configs are ignored, so any of the overrides done by other configs are not reflected.
- If you run `eslint .` inside of the nested dir, now the other config is loaded, so it _sometimes_ works, depending on the current working directory.
- All of the paths inherited from the root array will be relative to the wrong dir entirely, e.g. ignoring `scripts/**` in the root will also ignore `script/**` in the local dir, even if the intent was that only the top level dir was ignored. Or, maybe a rule applied to a completely different location and shouldn't be included at all.

The only way to do this is to forego multiple configs entirely. The root `eslint.config.js` could include the overrides for _everything_, or invent a custom config loading mechanism, like (in psuedocode):

```js
import { globSync } from "glob";

const allConfigs = globSync("**/eslint.config.json");

export default [
    // ...
    allConfigs.map((f) => {
        // Somehow read in the config, then remap all of its internal `files` globs to match a subdir?
    })
];
```

But, now you're loading more configs than may be needed for a given file. This is especially a problem for DefinitelyTyped; the _9244_ package monorepo currently has _2247_ `.eslintrc.json` files:

```console
$ find . -iname 'package.json' | wc -l
9244
$ find . -iname '.eslintrc.json' | wc -l
2247
```

Normally, this would make me quite worried because if you run eslint in parallel per package (like DT does), that could mean `9,244 * 2,247 = 2,0771,268` config loads. But, under flat config, each config is a module, so the runtime should theoretically be reusing that module, so the reality would likely be closer to just "load 2000 JSON files once" so long as the config is not invalidated.

That doesn't solve the problem of "what does this config file look like", however; not everything can be represented in JSON anymore with flat config.

### What do you think is the correct solution?

ESLint could look for the closest `eslint.config.js` file. This is similar to old configs, but explicitly _without_ the merging semantics. This would make ESLint working directory invariant; no matter where you run things, you get the same result.

I understand that there is some concern about the performance of this, but I don't actually think it will cost too much; you only need to walk a given dir once and remember which config applied to it. All in all, that's no more accesses to the disk than you already need to actually be able to list the files you're going to lint.

However, this does not fix the other flaw; inheriting another directory's config does not imply any modifications of file globs. If I spread in a parent dir's eslint config, none of the paths will line up, and I have no solution to that.

---

Another thing I thought of was to have/suggest a layout like:

```
.
├── eslint.base.config.js - contains the actual shared config
├── eslint.config.js      - imports shared config, all child dir configs
└── packages/
    ├── foo/
    │   └── eslint.config.js  - imports only shared config
    └── bar/
        └── eslint.config.js  - imports only shared config
```

This would also be cwd-invariant, but is pretty complicated, and also does not solve the problem of re-rooting paths from other configs / removing ones that don't apply to the subdir.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

I have yet to do perf testing on the "load 2247 config files" method or the "base config" idea; I've been really busy and have not had time to work on DT related stuff like this.

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Implemented the design from https://github.com/eslint/rfcs/pull/120:

- New `unstable_config_lookup_from_file` flag
- Refactored config lookup into `ConfigLoader` and `LegacyConfigLoader`
- Refactored `findFiles()` to use new config scheme.
- Updated documentation
- Updated tests

Fixes #18385

#### Is there anything you'd like reviewers to focus on?

Marking as a draft because I'd like us to merge https://github.com/eslint/eslint/pull/18134 first.

Note that I did not remove the config loading logic from the `ESLint` class, I just duplicated it in the `config-loader.js` file in order to have an easier merge once #18134 is merged. (This accounts for the linting error in `eslint.js`)

### CODE REVIEW

**Comment by fasttime** [comment]
There are merge conflicts now.

File: lib/config/config-loader.js [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+ * @throws {Error} If no configuration exists.
+ */
+function assertConfigurationExists(configFilePath, loaderOptions) {
+
+    const {
+        configFile: useConfigFile,
+        overrideConfig
+    } = loaderOptions;
+
+    if (!configFilePath && useConfigFile !== false && (!overrideConfig || overrideConfig[Symbol.for("eslint:from-cli")])) {
```
Comment (fasttime) [inline lib/config/config-loader.js:85]: If the last condition in this check is removed, all tests will still pass.

```suggestion
    if (!configFilePath && useConfigFile !== false) {
```

Is there still a reason for attaching a symbol to `overrideConfig` arrays from the CLI?

File: tests/lib/eslint/eslint.js [hunk 2]
```diff
@@ -143,8848 +147,9078 @@ describe("ESLint", () => {
...
-    describe("ESLint constructor function", () => {
+    [
+        [],
+        ["unstable_config_lookup_from_file"]
+    ].forEach(flags => {
 
-        it("should have a static property indicating the configType being used", () => {
-            assert.strictEqual(ESLint.configType, "flat");
-        });
+        describe("ESLint constructor function", () => {
```
Comment (fasttime) [inline tests/lib/eslint/eslint.js:155]: Probably it's not necessary to repeat every test in this suite with and without the `unstable_config_lookup_from_file` flag. I understand the caution but many of these unit tests are not related to the config lookup.

File: lib/eslint/eslint-helpers.js [hunk 3]
```diff
@@ -278,93 +257,107 @@ async function globSearch({
...
+                 *
+                 * This special case only applies to immediate children of `basePath`, so we
+                 * don't need to worry about any paths that don't include a slash.
+                 *
+                 * Basically, if the raw pattern isn't "." and the entry path doesn't include
+                 * a slash, then we don't need to look for a config file.
+                 */
+                const configs = rawPatterns[matchIndex] !== "." && !entry.path.includes("/")
+                    ? void 0
+                    : await configLoader.loadConfigArrayForDirectory(absolutePath);
```
Comment (mdjermanovic) [inline lib/eslint/eslint-helpers.js:305]: Special-casing only immediate children would cause inconsistent behavior, in my opinion. For example, if the root config file ignores `foo` and `bar/baz` directories, and both of them have their own config files, `eslint foo` would work, while `eslint bar/baz` wouldn't.

File: tests/lib/eslint/eslint.js (hunk 2 above)
Comment (nzakas) [reply tests/lib/eslint/eslint.js:155]: It's easier to do this than it is to try and identify which tests would actually be affected.

File: lib/config/config-loader.js (hunk 1 above)
Comment (nzakas) [reply lib/config/config-loader.js:85]: Hmmm, that is weird. There was definitely at least one test that was failing without this.

Basically, the CLI always passes `overrideConfig`, so I can't rely on it being `undefined` for this check.

File: lib/eslint/eslint-helpers.js (hunk 3 above)
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:305]: Sorry, I'm having trouble following what it is you're describing. Can you put together a test case that will fail?

And do you have other suggestions for how you might approach this? It turns out getting the exact behavior in the RFC is pretty difficult and this was the only approach I found that passed all of the tests we had described there.

File: lib/cli.js [hunk 4]
```diff
@@ -348,14 +350,9 @@ const cli = {
...
-            basePath,
-            error
+            basePath
         } = await locateConfigFileToUse({ cwd: process.cwd(), configFile }, hasUnstableTSConfigFlag);
 
-        if (error) {
-            throw error;
-        }
-
         return ["--config", configFilePath, "--basePath", basePath];
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/cli.js:356]: `locateConfigFileToUse` now throws an error.

File: lib/eslint/eslint.js [hunk 5]
```diff
@@ -710,6 +469,34 @@ class ESLint {
...
+            allowTS: processedOptions.flags.includes("unstable_ts_config")
+        };
+
+        this.#configLoader = this.hasFlag("unstable_config_lookup_from_file")
+            ? new ConfigLoader(configLoaderOptions)
+            : new LegacyConfigLoader(configLoaderOptions);
+
+        debug(`Using config loader ${this.#configLoader.constructor.name}`);
+
+        privateMembers.set(this, {
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint.js:489]: Thanks. Looks like a merge/rebase issue.

**Comment by fasttime** [comment]
There is a new merge conflict now and the tests were not run.

File: lib/eslint/eslint-helpers.js (hunk 3 above)
Comment (mdjermanovic) [reply lib/eslint/eslint-helpers.js:305]: Here's a test case: https://github.com/eslint/eslint/commit/bb439b137c4ee016934895f9f7ee866e7b671916

Comment (mdjermanovic) [reply lib/eslint/eslint-helpers.js:305]: One approach could be to search per each pattern and start loading configs from `globParent()` of the pattern. That's how eslintrc's FileEnumerator works. A downside is performance as the same directories might be traversed multiple times.

Another approach could be to keep searching with multiple patterns at once but check if the directory path is a part of the `globParent()` of at least one of the patterns, and in that case enter the directory even if it's ignored by a config in a parent directory.

File: lib/config/config-loader.js (hunk 1 above)
Comment (nzakas) [reply lib/config/config-loader.js:94]: Yeah. I actually tried refactoring so that the CLI would only pass `overrideConfig` if there were actual overrides, but it quickly turned into a giant mess. Perhaps with more refactoring around it, this can be cleaned up in the future.

**Comment by fasttime** [comment]
LGTM, thanks! Leaving open for @mdjermanovic to complete his review.

**Review by fasttime — APPROVED** [review]

File: lib/eslint/eslint-helpers.js [hunk 6]
```diff
@@ -277,94 +256,100 @@ async function globSearch({
...
+                /*
+                 * Special case: Don't look for a config file when "." is passed. This
+                 * is because we don't want to ignore the directory itself, just its
+                 * contents.
+                 */
+                const configs = rawPatterns[matchIndex] !== "."
+                    ? void 0
+                    : await configLoader.loadConfigArrayForDirectory(absolutePath);
+
+                return !configs?.isDirectoryIgnored(absolutePath);
```
Comment (mdjermanovic) [inline lib/eslint/eslint-helpers.js:299]: Now, on this branch, a pattern like `**/*.js` would traverse all directories, which is not what we want, I believe.

To reproduce, add `console.log(entry.path);` at the beginning of `async directoryFilter(entry)` and run `npx eslint "**/*.js"`.

```
.cache
.git
.git/hooks
.git/info
.git/logs
.git/logs/refs
.git/logs/refs/heads
.git/logs/refs/remotes
.git/logs/refs/remotes/origin
.git/logs/refs/remotes/origin/chore
.git/logs/refs/remotes/origin/docs
.git/logs/refs/remotes/origin/fix
.git/logs/refs/remotes/origin/renovate
.git/objects
.git/objects/00
.git/objects/01
.git/objects/02
.git/objects/03
.git/objects/04
.git/objects/05
.git/objects/06
.git/objects/07
.git/objects/08
.git/objects/09
.git/objects/0a
.git/objects/0b
.git/objects/0c
.git/objects/0d
.git/objects/0e
.git/objects/0f
.git/objects/10
.git/objects/11
.git/objects/12
.git/objects/13
.git/objects/14
.git/objects/15
.git/objects/16
.git/objects/17
.git/objects/18
.git/objects/19
.git/objects/1a
.git/objects/1b
.git/objects/1c
.git/objects/1d
.git/objects/1e
.git/objects/1f
.git/objects/20
.git/objects/21
.git/objects/22
.git/objects/23
.git/objects/24
.git/objects/25
.git/objects/26
.git/objects/27
.git/objects/28
.git/objects/29
.git/objects/2a
.git/objects/2b
.git/objects/2c
.git/objects/2d
.git/objects/2e
.git/objects/2f
.git/objects/30
.git/objects/31
.git/objects/32
.git/objects/33
.git/objects/34
.git/objects/35
.git/objects/36
.git/objects/37
.git/objects/38
.git/objects/39
.git/objects/3a
.git/objects/3b
.git/objects/3c
.git/objects/3d
.git/objects/3e
.git/objects/3f
.git/objects/40
.git/objects/41
.git/objects/42
.git/objects/43
.git/objects/44
.git/objects/45
.git/objects/46
.git/objects/47
.git/objects/48
.git/objects/49
.git/objects/4a
.git/objects/4b
.git/objects/4c
.git/objects/4d
.git/objects/4e
.git/objects/4f
.git/objects/50
.git/objects/51

... (~10,000 directories)
```

File: lib/eslint/eslint-helpers.js (hunk 3 above)
Comment (mdjermanovic) [reply lib/eslint/eslint-helpers.js]: Here's another test that I think should pass but is failing with the current iteration of this PR: https://github.com/eslint/eslint/commit/7e2193771c8b4f0452c3ec531743b8db0dbd2940

Comment (mdjermanovic) [reply lib/eslint/eslint-helpers.js]: Another possible approach: perhaps we could always group patterns by their `globParent()` and start loading configs from that path.

File: lib/eslint/eslint-helpers.js (hunk 6 above)
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:299]: Hmmm yeah, that's not right. I'll take a look.

Comment (nzakas) [reply lib/eslint/eslint-helpers.js:299]: I'm not able to reproduce this (Windows 10). Are you sure you're using the right command (`npx eslint` would download ESLint if you don't have it locally?). Here's what I'm getting:

```
$ node ./bin/eslint.js "**/*.js"
(node:5064) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. 
Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show 
where the warning was created)
.cache
.git
.github
.nyc_output
.trunk
.vscode
bin
build
conf
coverage
docs
lib
messages
node_modules
packages
templates
tests
tmp
tools

Oops! Something went wrong! :(

ESLint: 9.9.1

No files matching the pattern "**C:/Program Files/Git/*.js" were found.
Please check for typing mistakes in the pattern.
```

File: tests/lib/eslint/eslint.js [hunk 7]
```diff
@@ -143,8848 +147,9092 @@ describe("ESLint", () => {
...
+                    cwd: path.join(fixtureDir, ".."),
+                    overrideConfigFile: getFixturePath("eslint.config.js"),
+                    ignore: false,
+                    overrideConfig: {
+                        rules: {
+                            semi: 2
+                        }
+                    }
+                });
+                const failFilePath = fs.realpathSync(ge
```
[the remark this answers is not in the document]
Comment (nzakas) [reply tests/lib/eslint/eslint.js:9063]: I don't think we explicitly defined this behavior. Technically, throwing an error is correct because our logic says "for a given file, if you can't find a config file that tells you what to do with it, throw an error."

I'm not sure how common it would be for someone to run `eslint .` in a directory without a config file. This might be a case where we leave this behavior as-is and see what feedback we get. Otherwise, I'd be back to checking "if pattern is `.` and file is an immediate child of `.`, then ignore it".

File: lib/eslint/eslint-helpers.js (hunk 6 above)
Comment (mdjermanovic) [reply lib/eslint/eslint-helpers.js:299]: > Oops! Something went wrong! :(
> 
> ESLint: 9.9.1
> 
> No files matching the pattern "**C:/Program Files/Git/*.js" were found.
> Please check for typing mistakes in the pattern.

That `"**C:/Program Files/Git/*.js"` looks weird, I'm not sure where it did come from when running `$ node ./bin/eslint.js "**/*.js"` (maybe your shell is doing some expansion despite the quotes?).

I tried again, on this branch, with a clean `npm install`, and in cmd.exe which doesn't expand globs. I also added `console.log({ patterns });` at the beginning of `async function globSearch` to make sure that the pattern is passed through.

```js
C:\projects\eslint>node ./bin/eslint.js "**/*.js"
(node:7136) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show where the warning was created)

{ patterns: [ 'C:\\projects\\eslint\\**\\*.js' ] }
.cache
.git
.git/hooks
.git/info
.git/logs
.git/logs/refs
.git/logs/refs/heads
.git/logs/refs/remotes
.git/logs/refs/remotes/origin
.git/logs/refs/remotes/origin/chore
.git/logs/refs/remotes/origin/docs
.git/logs/refs/remotes/origin/fix
.git/logs/refs/remotes/origin/renovate
.git/objects
.git/objects/00
.git/objects/01
.git/objects/02
.git/objects/03
.git/objects/04
.git/objects/05
.git/objects/06
.git/objects/07
.git/objects/08
.git/objects/09
.git/objects/0a
.git/objects/0b
.git/objects/0c

...
```

Comment (nzakas) [reply lib/eslint/eslint-helpers.js:299]: Okay, I can see this in `cmd` but not in Git Bash. It looks like Git Bash is representing the path in Windows differently:

```
  patterns: [
    'C:\\Users\\nzaka\\projects\\eslint\\eslint\\**C:\\Program Files\\Git\\*.js'  
  ]
```

I'll take a look at that, too.

Comment (nzakas) [reply lib/eslint/eslint-helpers.js:299]: So it looks like Git for Windows is expanding the pattern even if it's quoted.
https://github.com/git-for-windows/git/issues/5131

**Comment by nzakas** [comment]
@mdjermanovic I've been fighting with this for the last three days and have made zero progress. Do you want to take a stab at it?

**Comment by nzakas** [comment]
Hold that thought! I might have figured it out.

File: lib/eslint/eslint-helpers.js [hunk 8]
```diff
@@ -277,94 +256,157 @@ async function globSearch({
...
+                 * way that the pattern can match `entry.path` yet.
+                 *
+                 * 4. A directory glob pattern is passed (e.g., "subd*"). We want
+                 * this case to act like case 1, where we don't look for a config
+                 * file in the parent directory.
+                 *
+                 * 5. A recursive glob pattern is passed (e.g., "**"). We want this
+                 * case to act like case 2.
+                 */
+                const matchedPattern = matchers[matchIndex].pattern;
```
Comment (mdjermanovic) [inline lib/eslint/eslint-helpers.js:332]: I think the order of patterns shouldn't matter. But since we are using only the first match to determine whether to traverse into a subdirectory, the order would matter.

For example, if root config file ignores `subdir`, and `subdir` has it's own config file that doesn't ignore anything, when we run:

* `eslint "subdir/*.js" "**/*.cjs" --flag unstable_config_lookup_from_file`: it will enter `subdir`, and even lint .cjs files inside it along with .js files.
* `eslint "**/*.cjs" "subdir/*.js" --flag unstable_config_lookup_from_file`: it won't enter `subdir`, and will throw an error for unmatched pattern `subdir/*.js`.

File: lib/eslint/eslint-helpers.js [hunk 9]
```diff
@@ -277,94 +256,157 @@ async function globSearch({
...
+                 *
+                 * 3. A directory is passed in the form of "subdir/subsubdir".
+                 * In this case, we don't want to look for a config file in the
+                 * parent directory ("subdir"). We can skip looking for a config
+                 * file until `entry.depth` is greater than 1 because there's no
+                 * way that the pattern can match `entry.path` yet.
+                 *
+                 * 4. A directory glob pattern is passed (e.g., "subd*"). We want
+                 * this case to act like case 1, where we don't look for a config
+                 * file in the parent directory.
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:327]: 🤯 I'll adjust this.

File: lib/eslint/eslint-helpers.js (hunk 8 above)
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:332]: Hmm, good point. I'll look into this.

Comment (nzakas) [reply lib/eslint/eslint-helpers.js:333]: For clarity, the correct behavior is the second one: we should throw an error when any pattern has zero results. This matches current behavior.

File: lib/eslint/eslint-helpers.js [hunk 10]
```diff
@@ -277,94 +256,168 @@ async function globSearch({
...
+                 * way that the pattern can match `entry.path` yet.
+                 *
+                 * 4. A directory glob pattern is passed (e.g., "subd*"). We want
+                 * this case to act like case 2 because it's unclear whether or not
+                 * any particular directory is meant to be traversed.
+                 *
+                 * 5. A recursive glob pattern is passed (e.g., "**"). We want this
+                 * case to act like case 2.
+                 */
+                const shouldLoadConfigFile = matchedPatterns.some(pattern => {
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:329]: > Did you maybe consider the approach of always grouping patterns by their glob parent in `findFiles()`?

I didn't understand what you meant by this, so I didn't consider it. If you'd like to explain more, or maybe write up an example, I'm happy to.

I do think we're starting to get into edge cases, though, and it might be better to merge the PR and then tackle edge cases individually. That's part of the benefit of having this behind a flag.

File: lib/eslint/eslint-helpers.js [hunk 11]
```diff
@@ -277,94 +256,168 @@ async function globSearch({
...
-                    reject(error);
-                } else {
-                    resolve(entries);
+                // entries may be directories or files so filter out directories
+                if (entry.isDirectory) {
+                    return false;
                 }
+
+                const configs = await configLoader.loadConfigArrayForFile(absolutePath);
+                const config = configs.getConfig(absolutePath);
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:378]: Hmm, maybe not. In the original file, there was a relative path that was normalized to Posix, but maybe the absolute path doesn't need it.

File: lib/config/config-loader.js [hunk 12]
```diff
@@ -0,0 +1,729 @@
...
+        throw new Error("'filePath' must be a non-empty string");
+    }
+}
+
+/**
+ * Asserts that a configuration exists. A configuration exists if any
+ * of the following are true:
+ * - `configFilePath` is defined.
+ * - `useConfigFile` is defined.
+ * - `overrideConfig` is defined.
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/config/config-loader.js:72]: All the current tests are passing, so I think the comment might be wrong and the code is correct. The condition below seems to match your expectations, but if you can come up with a test that fails, let me know.

[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/config/config-loader.js:72]: Here are three tests that pass on the `main` branch but fail (both with and without the `unstable_config_lookup_from_file`flag) on this branch : https://github.com/eslint/eslint/commit/5f22aed9896696787f60fd9c0e522b08f4f8ff9d. The commit is made on top of this branch.

File: lib/eslint/eslint-helpers.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/eslint/eslint-helpers.js:329]: I was referring to this code:

https://github.com/eslint/eslint/blob/b0df2e371fa8fd097cb31d914181a54df58e9e7e/lib/eslint/eslint-helpers.js#L603-L642

Patterns inside `cwd` are in the same group, while the rest are grouped by `globParent(pattern)`. If we always group patterns by `globParent(pattern)`, then we'll have groups such that for all patterns in the group the same configs apply.

In the example from my comment above, `src/*.js` and `**/*.cjs` would be in different groups.

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:329]: Ah! Thank you for explaining. That makes sense. Let me give that a try.

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:329]: That does work for a lot of the cases, but for `npx eslint src`, it now throws an error saying it's unmatched. So, it seems like we'll still need additional logic. Just three failing tests, though, so it's close.

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/eslint-helpers.js:329]: Okay, I think this piece is working now. :tada:

File: lib/config/config-loader.js [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+            const relativeIgnorePath = path.relative(basePath, cwd).replace(/\\/gu, "/");
+
+            relativeIgnorePatterns = ignorePatterns.map(pattern => {
+                const negated = pattern.startsWith("!");
+                const basePattern = negated ? pattern.slice(1) : pattern;
+
+                return (negated ? "!" : "") +
+                    path.posix.join(relativeIgnorePath, basePattern);
+            });
+        }
```
Comment (mdjermanovic) [inline lib/config/config-loader.js:273]: This doesn't work well if `cwd` is not inside `basePath` (which can be the case with the new flag) as the calculated patterns will start with `../` and so will be effectively ignored. But it seems quite complex to fix, so maybe we can leave it for a follow-up PR.

File: lib/eslint/eslint-helpers.js [hunk 14]
```diff
@@ -569,17 +602,17 @@ async function findFiles({
...
-                ({ patterns: globbyPatterns, rawPatterns } = searches.get(cwd));
-            } else {
-                if (!searches.has(basePath)) {
-                    searches.set(basePath, { patterns: [], rawPatterns: [] });
-                }
-                ({ patterns: globbyPatterns, rawPatterns } = searches.get(basePath));
+            if (!searches.has(basePath)) {
+                searches.set(basePath, { patterns: [], rawPatterns: [] });
             }
+            ({ patterns: globbyPatterns, rawPatterns } = searches.get(basePath));
```
Comment (mdjermanovic) [inline lib/eslint/eslint-helpers.js:615]: The same change should be done for directories (lines 585-593) as they work the same as globs.

File: lib/eslint/eslint-helpers.js [hunk 15]
```diff
@@ -277,94 +259,145 @@ async function globSearch({
...
-                if (!promiseRejected) {
-                    try {
-                        return filter(...args);
-                    } catch (error) {
-                        promiseRejected = true;
-                        reject(error);
+    const walk = hfs.walk(
+        basePath,
+        {
+            async directoryFilter(entry) {
```
Comment (mdjermanovic) [inline lib/eslint/eslint-helpers.js:267]: Then, I believe this can be simplified to something like:

```js
async directoryFilter(entry) {

    if (!matchers.some(matcher => matcher.match(entry.path, true))) {
        return false;
    }

    const absolutePath = path.resolve(basePath, entry.path);
    const configs = await configLoader.loadConfigArrayForDirectory(absolutePath);

    return !configs.isDirectoryIgnored(absolutePath);
},
```

**Comment by sam3k** [comment]
Per TSC meeting on 2024-09-19, this is a high priority pull request and the TSC team will work together to get it to the finish line as soon as possible.

File: lib/config/config-loader.js (hunk 13 above)
Comment (mdjermanovic) [reply lib/config/config-loader.js:273]: Here's the issue: https://github.com/eslint/eslint/issues/18948

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!
